[Core]Enhancement MPC  builder and solver

adityaghantasala opened the pull request:
Important points in this PR : 

- Adds process info to the steps of the master-slave constraint class (@loumalouomega  as discussed in #2942 ) and corresponding changes.

- Removing shrink_to_fit function calls on vectors at places to save memory allocation time. 

- @RiccardoRossi  I implemented a better version for doing operation K = T'*K*T and F = T'*(F - K*g). When you were here we discussed some possible bottlenecks. Now I implemented a in-place modification of the K matrix to achieve this. This version is way faster than the previous implementation.

- changes to loop for assembling the global m-s constraints.  

- Checking the size of mGlobalMasterSlaveConstraints to switch between the constraint and non constraint version (instead of number of constraints on the modelpart)

adityaghantasala committed:
using global ms constraint map to check for constrints presense

adityaghantasala committed:
removing shrink to fit to save time

adityaghantasala committed:
adding function for inplace matrix modification and required modifications

adityaghantasala committed:
updating comments

adityaghantasala committed:
removing un used functions

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into core-enhancement-mpc-b-and-s

adityaghantasala committed:
adding process info to base ms constraint steps

adityaghantasala committed:
adding process info when calling base class functions in b and s with constraints

adityaghantasala committed:
splitting the modification into lhs and rhs functions

adityaghantasala committed:
removing parallel assembly of the constraints

adityaghantasala committed:
removing residual shrink to fit

adityaghantasala committed:
changes to openmp loop to avoid potential segfaults

loumalouomega commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -278,33 +278,33 @@ class MasterSlaveConstraint
      * @brief This is called in the beginning of each solution step
      */
-    virtual void InitializeSolutionStep()
+    virtual void InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo)
```
:D

loumalouomega commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -198,7 +198,7 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         {
             auto it = constraints_begin + k;
-            it->InitializeSolutionStep(); // Here each constraint constructs and stores its T and C matrices. Also its equation ids.
+            it->InitializeSolutionStep(rModelPart.GetProcessInfo()); // Here each constraint constructs and stores its T and C matrices. Also its equation ids.
```
Better to move this outside the loop, you are accessing to the process info everytime

loumalouomega commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -219,7 +219,7 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         {
             auto it = constraints_begin + k;
-            it->FinalizeSolutionStep();
+            it->FinalizeSolutionStep(rModelPart.GetProcessInfo());
```
Idem

loumalouomega committed:
Moving process info from loop

loumalouomega reviewed (dismissed):
OK

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -602,11 +607,9 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
-                it->CalculateLocalSystem(relation_matrix, constant_vector, r_current_process_info);
                 //assemble the Constraint contribution
-                AssembleSlaves(slave_equation_ids, master_equation_ids, relation_matrix);
+                #pragma omp critical
```
why critical? this kills the parallelism...

adityaghantasala replied:
the critical is necessary for synchronized access to the map `mGlobalMasterSlaveConstraints`, which is a map of id and a unique pointer. If there is no `critical` there, there will be random segfaults.

RiccardoRossi replied:
yes i assumed something of the sort :-). The point is: this critical is there only in the construction phase or any time you do build?

RiccardoRossi commented on kratos/utilities/helper_classes_for_constraint_builder.h:
```diff
@@ -605,84 +570,137 @@ class ConstraintImposer {
+            IndexType master_i_other = 0;
+            for (auto& master_index_other : mLocalIndices.processed_master_indices) {
+                rLHSContribution(master_index, master_index_other) += mLocalIndices.container_master_weights[master_i] *
+                                                                        rLHSContribution(mLocalIndices.container_master_slaves[master_i], mLocalIndices.container_master_slaves[master_i_other])
```
are u accessing to a map here? take care is expensive...

adityaghantasala replied:
These are vectors @RiccardoRossi  not maps.

RiccardoRossi commented on kratos/utilities/helper_classes_for_constraint_builder.h:
```diff
@@ -605,84 +570,137 @@ class ConstraintImposer {
+                //master_weight = mTransformationMatrixLocal(slave_index,master_index);
+                master_weight = master_weights_vector(i_master);
+                for (auto& internal_index : mLocalIndices.internal_index_vector) {
+                    rRHSContribution(internal_index) += -rLHSContribution(internal_index, slave_index) * slave_constant;
```
-=

RiccardoRossi reviewed (commented):
minor comments...

loumalouomega committed:
Adding Check

loumalouomega reviewed (dismissed):
@adityaghantasala I added a Check method, could be relevant and important

philbucher commented:
+1 for Check!

adityaghantasala commented:
@loumalouomega  +1

adityaghantasala committed:
calling constraints check in model part check

adityaghantasala committed:
-= change

adityaghantasala committed:
removing un used array declaration

adityaghantasala committed:
missing reference

adityaghantasala committed:
adding resize of new vectors in localIndices

loumalouomega reviewed (dismissed):
Nice

adityaghantasala commented:
@loumalouomega  I changed ModelPart.cpp to call the check of constraint class along with element and condition.  
Also addressed comments of @RiccardoRossi  to change to -=

philbucher commented on kratos/sources/model_part.cpp:
```diff
@@ -1518,6 +1518,9 @@ int ModelPart::Check(ProcessInfo& rCurrentProcessInfo) const
         err = condition_iterator->Check(rCurrentProcessInfo);
+    for (MasterSlaveConstraintConstantIteratorType constraint_iterator = MasterSlaveConstraintsBegin(); 
+            constraint_iterator != MasterSlaveConstraintsEnd(); constraint_iterator++)
+        err = constraint_iterator->Check(rCurrentProcessInfo);
```
hm won't this just return the value of the last constraint that was checked?
I guess this is also wrong in the current master

adityaghantasala replied:
it does.  will suggest to leave it here and change it a different PR.

loumalouomega committed:
Unnecessary virtual

adityaghantasala commented:
@RiccardoRossi  shall I proceed and merge this ? any more comments ?

philbucher commented:
@adityaghantasala I just saw that you are not passing the ProcessInfo by const-reference, only by reference
Can you please change it to const-reference?
Changing it afterwards will be painful

Edit you should also pass it to `Initialize` and `Finalize`

See discussions in #1897 and #2993 to pass ProcessInfo by const-ref to Conditions and Elements

philbucher commented:
Also `GetDofList` and `EquationIdVector` should be const functions, see #732

adityaghantasala committed:
cont process info and functions in base and line m-s constraint

adityaghantasala commented:
Done !

philbucher commented:
can you please pass the processinfo also to Initialize and Finalize? => #2817

adityaghantasala committed:
process into to initalize and finalize

adityaghantasala committed:
Merge branch 'core-enhancement-mpc-b-and-s' of https://github.com/KratosMultiphysics/Kratos into core-enhancement-mpc-b-and-s

adityaghantasala commented:
sorry missed it .. added it ..

philbucher reviewed (dismissed):
Thanks!

One question: Aren't you calling initialize & finalize somwhere? if so, then also there the ProcessInfo should be passed

adityaghantasala commented:
good question .. I am not calling them ... but they should be called in scheme's initialize function (elements and conditions are there). we can change that in the next pr.

philbucher commented:
ok
All good from my side then

loumalouomega commented:
I was speaking with @RiccardoRossi , and not for this PR but for a future one. It would be interesting to integrate inside the MasterSlavConstraint its own EquationIdVector method

adityaghantasala commented:
@loumalouomega  I cannot see the application now. But yes .. it is possible to do it like that.

adityaghantasala commented:
I am proceeding with merging this. the suggested changes to modelpart.cpp will be done in a different PR.

loumalouomega commented:
The application would be to add equation id of dof not added directly by other conditions or elements

adityaghantasala commented:
Then it is also important that it reflects in GetDofsVector function.

loumalouomega commented:
Of course, I meant the couple

msandre commented:
@adityaghantasala can you wait until start of next week so I can take a look?

RiccardoRossi commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -359,6 +359,25 @@ class MasterSlaveConstraint
+    {
+        KRATOS_TRY
+
+        KRATOS_ERROR_IF( this->Id() < 1 ) << "MasterSlaveConstraint found with Id " << this->Id() << std::endl;
```
this only throws if the id is zero no?

adityaghantasala replied:
is 0 and less than zero or ?

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -213,13 +218,18 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         BaseType::FinalizeSolutionStep(rModelPart, A, Dx, b);
+	
+	// Getting process info
+	ProcessInfo& r_process_info = rModelPart.GetProcessInfo();
```
const ProcessInfo& ?

RiccardoRossi reviewed (commented):
in general i am fine with this. just few minor comments

msandre commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -192,13 +192,18 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         BaseType::InitializeSolutionStep(rModelPart, A, Dx, b);
+	
+	// Getting process info
+	ProcessInfo& r_process_info = rModelPart.GetProcessInfo();
```
const

msandre commented on kratos/utilities/helper_classes_for_constraint_builder.h:
```diff
@@ -605,84 +572,137 @@ class ConstraintImposer {
-    void CalculateLocalTransformationMatrix(LocalIndicesType& rLocalIndices,
-                                            MatrixType& rTransformationMatrixLocal,
-                                            EquationIdVectorType& rEquationIds)
+    void ModifyLHS(MatrixType &rLHSContribution, VectorType& rRHSContribution, EquationIdVectorType &rEquationIds)
```
Could you make the function name more descriptive?

msandre reviewed (commented):
Some minor comments from my side. When things have settled, it would be good to add a wiki describing the overall class design and how it interfaces with the other core classes.

adityaghantasala commented:
@RiccardoRossi  yes the critical is only in construction phase (initialization) and **does not** happen every time we build (of course then the constraint relations should stay constant)

adityaghantasala commented:
@msandre  I addressed your comments also

adityaghantasala commented:
sorry for noise .. I committed some local files.

adityaghantasala committed:
review changes

adityaghantasala committed:
function name changes

loumalouomega commented:
@adityaghantasala you forgot to remove the branch
